keps/sig-api-machinery/20190226-network-proxy.md

```diff
@@ -0,0 +1,618 @@
+- **Flat Network** A network where traffic can be successfully routed using IP. Implies that there are no overlapping (i.e. shared) IPs on the network.
+
+## Proposal
+
+The proposal is to run connectivity proxy inside master network. 
+The connectivity proxy exposes a gRPC interface to API server. 
+Each connectivity proxy allows secure connections to one or more cluster networks. 
+Any network addressed by a connectivity proxy must be flat. 
+Currently the only mechanism for handling overlapping IP ranges in Kubernetes is the Proxy. 
+As we are passed the proxy there would need to be a non Kubernetes mechanism 
+for routing that traffic properly.
+
+Running the connectivity proxy in a separate process has a few advantages. 
+It allows for easy extension of the connectivity proxy without needing to recompile the KAS. 
+In conjunction with that, it means the administrators of the cluster can run their own variants 
+of the connectivity proxy. 
+Creating a connectivity proxy which audits traffic for debuggability or forwards traffic via a 
+proprietary VPN is now considerably easier to accomplish. 
+It also lowers the complexity of the KAS code base. 
+It separates the concerns of master <-> cluster connectivity from serving the Kubernetes API. 
+Separating these concerns reduces the effects of an issue. 
+Connectivity issues will not stop the KAS from serving API requests. 
+This is important as serving API requests may necessary in order to fix the problem. 
+For instance, if there is a problem with a node, set of nodes or load-balancers configuration, 
+then we need to make API requests to fix that problem. 
+If connectivity is issue is crashing the process its in and the process it is in is the KAS, 
+then we have a problem. 
+We need to make API requests to fix the issue but the KAS which serves API requests is down.  
+
+![Network Proxy Simple Cluster](NetworkProxySimpleCluster.png)
+The diagram shows API Server’s outgoing traffic flow. 
+In this diagram, we have user (in blue box), master network (in purple cloud) and 
+a cluster network (in green cloud). 
+
+The user (blue) initiates communications to API Server. 
+As part of serving the request from the user, 
+API Server needs to initiate connections to other components. 
+It could be node/pod/service in cluster networks (red dotted arrow to green cloud), 
+or etcd for storage in the same master network (blue arrow) or mutate the request 
+based on an admission webhook (red dotted arrow to purple cloud). 
+API Server also needs to support NetworkContext based traffic routing, 
+to differentiate the above cases. 
+Connectivity Proxy should be able to do routing solely based on IP 
+without any additional NetworkContext, meaning the service CIDR, 
+node CIDR and pod CIDR of each cluster network cannot overlap. 
```

Comment:
That's a downer - I thought we were thinking of relaxing this with multiple proxies or contexts?